p2p/host/base/base_host.go

```diff
@@ -0,0 +1,255 @@
+	if err != nil {
+		return nil, err
+	}
+	n.Notify(newPeerConnectWatcher(h.emitters.evtPeerConnectednessChanged))
+
+	if err := h.initSignedRecord(); err != nil {
+		return nil, err
+	}
+	n.SetStreamHandler(h.newStreamHandler)
+	return h, nil
+}
+
+func (h *BaseHost) initSignedRecord() error {
+	cab, ok := peerstore.GetCertifiedAddrBook(h.network.Peerstore())
+	if !ok {
+		return errors.New("peerstore does not support signed records")
+	}
+	rec := peer.PeerRecordFromAddrInfo(peer.AddrInfo{ID: h.ID(), Addrs: h.Addrs()})
+	ev, err := record.Seal(rec, h.Peerstore().PrivKey(h.ID()))
+	if err != nil {
+		return fmt.Errorf("failed to create signed record for self: %s", err)
+	}
+	if _, err := cab.ConsumePeerRecord(ev, peerstore.PermanentAddrTTL); err != nil {
+		return fmt.Errorf("failed to persist signed record for self: %s", err)
+	}
+	return nil
+}
+
+func (h *BaseHost) Start() {
+	h.psManager.Start()
+}
+
+func (h *BaseHost) Addrs() []ma.Multiaddr {
+	addrs, err := h.Network().InterfaceListenAddresses()
+	if err != nil {
+		log.Debug("error retrieving network interface addrs: ", err)
+		return nil
+	}
+	return addrs
+}
+
+func (h *BaseHost) Connect(ctx context.Context, ai peer.AddrInfo) error {
+	// absorb addresses into peerstore
+	h.Peerstore().AddAddrs(ai.ID, ai.Addrs, peerstore.TempAddrTTL)
+
+	cs := h.Network().ConnsToPeer(ai.ID)
+	if len(cs) > 0 {
+		return nil
+	}
+
+	_, err := h.Network().DialPeer(ctx, ai.ID)
+	return err
+}
+
+func (h *BaseHost) ConnManager() connmgr.ConnManager {
+	return h.connmgr
+}
+
+func (h *BaseHost) Close() error {
+	h.closeOnce.Do(func() {
+		h.Network().Close()
+		h.connmgr.Close()
+		h.emitters.evtLocalProtocolsUpdated.Close()
+		h.emitters.evtPeerConnectednessChanged.Close()
+
+		h.psManager.Close()
+		if h.Peerstore() != nil {
+			h.Peerstore().Close()
+		}
+	})
+	return nil
+}
+
+func (h *BaseHost) EventBus() event.Bus {
+	return h.eventbus
+}
+
+func (h *BaseHost) Peerstore() peerstore.Peerstore {
+	return h.Network().Peerstore()
+}
+
+// Network returns the Network interface of the Host
+func (h *BaseHost) Network() network.Network {
+	return h.network
+}
+
+// Mux returns the Mux multiplexing incoming streams to protocol handlers
+func (h *BaseHost) Mux() protocol.Switch {
+	return h.mux
+}
+
+func (h *BaseHost) ID() peer.ID {
+	return h.Network().LocalPeer()
+}
+
+// newStreamHandler is the remote-opened stream handler for network.Network
+func (h *BaseHost) newStreamHandler(s network.Stream) {
+	protoID, handle, err := h.Mux().Negotiate(s)
+	if err != nil {
+		log.Infow("protocol negotiation failed", "error", err)
+		s.Reset()
+		return
+	}
+
+	s.SetProtocol(protocol.ID(protoID))
+	go handle(protoID, s)
+}
+
+func (h *BaseHost) NewStream(ctx context.Context, p peer.ID, protos ...protocol.ID) (network.Stream, error) {
+	s, err := h.Network().NewStream(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	selected, err := msmux.SelectOneOf(protocol.ConvertToStrings(protos), s)
```

Comment:
This won't obey the context unless we play the goroutine game (same as in BasicHost.NewStream).

Reply:
Fixed.